pkg/logql/log/parser.go

```diff
@@ -164,6 +188,49 @@ func (j *JSONParser) parseLabelValue(iter *jsoniter.Iterator, prefix, field stri
 
 func (j *JSONParser) RequiredLabelNames() []string { return []string{} }
 
+func (j *JSONParser) fastParseJSON(line []byte, lbs *LabelsBuilder) ([]byte, bool) {
+	for _, requiredKey := range j.RequiredJSONLabels {
+		jsonLine := string(line)
+		lineLens := len(jsonLine)
+		key := "\"" + requiredKey + "\":"
+		var beginIndex int
+		if j.tailIndexDirection {
+			beginIndex = strings.LastIndex(jsonLine, key)
+			if beginIndex < lineLens/2 {
+				j.tailIndexDirection = false
+			}
+		} else {
+			beginIndex = strings.Index(jsonLine, key)
+			if beginIndex > lineLens/2 {
+				j.tailIndexDirection = true
+			}
+		}
+		if beginIndex < 0 {
+			continue
+		}
+		subJSONLine := jsonLine[beginIndex:]
+		endIndex := strings.Index(subJSONLine, ",")
+		if endIndex < 0 {
+			endIndex = strings.Index(subJSONLine, "}")
+			if endIndex < 0 {
+				continue
+			}
+		}
+
+		value := jsonLine[beginIndex+len(key)+1 : beginIndex+endIndex]
+		valLen := len(value)
+		if valLen > 0 && strings.HasPrefix(value, "\"") {
+			// case:string
+			valStr := value[1 : valLen-1]
+			lbs.Set(requiredKey, valStr)
+		} else {
+			lbs.Set(requiredKey, value)
+		}
+	}
+
+	return line, true
+}
```

Comment:
There's some json library that I've been looking at that does this in a more clean way.

For example the each object of https://github.com/buger/jsonparser

Reply:
see https://github.com/grafana/loki/issues/4328